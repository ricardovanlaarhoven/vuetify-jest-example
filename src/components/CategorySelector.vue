<template>
    <div class="category-selector">
        <VAutocomplete
            label="Zichtbaarheid"
            :items="categories"
            item-text="name"
            item-value="id"
            v-model="visibilityCategoryId"
            :rules="[(v) => !!v || 'Zichtbaarheid is een verplicht veld']"
        />
        <VAutocomplete
            label="Hoofdcategorie"
            v-if="visibilityCategoryId"
            :items="parentCategories"
            item-text="name"
            item-value="id"
            v-model="parentCategoryId"
            :rules="[(v) => !!v || 'Hoofdcategorie is een verplicht veld']"
        />
        <VAutocomplete
            label="Subcategorie"
            v-if="parentCategoryId"
            :items="subCategories"
            item-text="name"
            item-value="id"
            v-model="categoryId"
            :rules="[(v) => !!v || 'Subcategorie is een verplicht veld']"
        />
    </div>
</template>

<script>
import {mapActions, mapGetters, mapState} from 'vuex';

export default {
    name: 'CategorySelector',
    props: {
        value: {
            type: Number,
        },
    },
    data() {
        return {
            categoryId: null,
            parentCategoryId: null,
            visibilityCategoryId: null,
        };
    },
    watch: {
        value: 'updateCategories',
        categoryId: 'emitCategory',
    },
    computed: {
        ...mapGetters({flattenCategories: 'DocumentCategories/flattenCategories'}),
        ...mapState('DocumentCategories', ['categories']),
        parentCategories() {
            if (!this.visibilityCategoryId) return;
            return this.categories.find(category => category.id === this.visibilityCategoryId).children;
        },
        subCategories() {
            if (!this.parentCategories) return;
            if (!this.parentCategoryId) return;
            return this.parentCategories.find(category => category.id === this.parentCategoryId).children;
        },
    },
    async created() {
        await this.getAllCategories();
        this.updateCategories(this.value);
    },
    methods: {
        ...mapActions({getAllCategories: 'DocumentCategories/getAll'}),
        /**
         * Set the visibilityCategory and parentCategory based on the categoryId
         */
        updateCategories(categoryId) {
            if (!categoryId) return;
            const category = this.flattenCategories.find(category => category.id === categoryId);
            if (!category.isFinalDepth && !category.parentId) {
                /* The selected category is a visibility category */
                this.visibilityCategoryId = categoryId;
            } else if (!category.isFinalDepth) {
                /* The selected category is a parentCategory */
                this.visibilityCategoryId = this.flattenCategories.find(category => category.id === categoryId).parentId;
                this.parentCategoryId = categoryId;
            } else {
                /* the selected category is the final category */
                this.categoryId = categoryId;
                this.parentCategoryId = category.parentId;
                this.visibilityCategoryId = this.flattenCategories.find(category => category.id === this.parentCategoryId).parentId;
            }
        },
        emitCategory(category) {
            this.$emit('input', category);
        },
    },
};
</script>

<style scoped>

</style>
