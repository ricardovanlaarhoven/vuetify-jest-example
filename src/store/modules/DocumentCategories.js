import getAllCategories from '../../api/documentCategories/getAll.js';

export default {
    namespaced: true,
    state: {
        categories: null,
    }, mutations: {
        set(state, categories) {
            state.categories = categories;
        },
    },
    getters: {
        flattenCategories: (state) => {
            if (!state.categories) return;
            /**
             * Flatten a multidimansional array
             * We need this method to obtain the parentId of a categoryId.
             * This categoryId could be in any level
             */
            const flatten = (children, currList = []) => {
                currList.push(...children.map(child => {
                    return {
                        id: child.id,
                        name: child.name,
                        parentId: child.parent_id,
                        isFinalDepth: child.is_final_depth,
                    };
                }));
                children.forEach(child => {
                    if (child.children === null) return;
                    flatten(child.children, currList);
                });
                return currList;
            };
            return flatten(state.categories);
        },
    },
    actions: {
        async getAll(context) {
            if (context.state.categories) {
                return;
            }
            const categories = await getAllCategories();
            context.commit('set', categories);
        },
        async invalidate(context) {
            context.commit('set', null);
            await context.dispatch('getAll');
        }
    },
};
