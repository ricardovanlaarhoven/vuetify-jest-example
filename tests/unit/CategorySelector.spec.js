import {createLocalVue, mount, shallowMount} from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Vuetify from 'vuetify';
import Vuex from 'vuex';
import CategorySelector from '../../src/components/CategorySelector.vue';
import DocumentCategories from '../../src/store/modules/DocumentCategories.js';

jest.mock('./../../src/api/documentCategories/getAll');
const localVue = createLocalVue();
describe('GlobalSearch.vue', () => {
    let vuetify;
    let store;

    beforeEach(() => {
        vuetify = new Vuetify();
        store = new Vuex.Store({
            modules: {DocumentCategories},
        });
    });

    it('matches ui', async () => {
        const wrapper = mount(CategorySelector, {
            localVue,
            vuetify,
            store,
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches ui when a headcategory', async () => {
        const wrapper = mount(CategorySelector, {
            localVue,
            vuetify,
            store,
            propsData: {
                value: 9,
            },

        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches ui when a subcategory is set', async () => {
        const wrapper = mount(CategorySelector, {
            localVue,
            vuetify,
            store,
            propsData: {
                value: 17,
            },
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches ui when using shallowMount', async () => {
        const wrapper = shallowMount(CategorySelector, {
            localVue,
            vuetify,
            store,
            propsData: {
                value: 17,
            },
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
