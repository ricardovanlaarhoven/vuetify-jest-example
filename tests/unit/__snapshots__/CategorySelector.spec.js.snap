// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalSearch.vue matches ui 1`] = `
<div class="category-selector">
  <div class="v-input theme--light v-text-field v-select v-autocomplete">
    <div class="v-input__control">
      <div role="combobox"
           aria-haspopup="listbox"
           aria-expanded="false"
           aria-owns="list-5"
           class="v-input__slot"
      >
        <div class="v-select__slot">
          <label for="input-5"
                 class="v-label theme--light"
                 style="left: 0px; position: absolute;"
          >
            Zichtbaarheid
          </label>
          <input id="input-5"
                 type="text"
                 autocomplete="off"
          >
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-menu-down theme--light"
              >
              </i>
            </div>
          </div>
          <input type="hidden">
        </div>
        <div class="v-menu">
          <div class="v-menu__content theme--light v-autocomplete__content"
               style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          >
          </div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper">
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GlobalSearch.vue matches ui when a headcategory 1`] = `
<div class="category-selector">
  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-select v-autocomplete">
    <div class="v-input__control">
      <div role="combobox"
           aria-haspopup="listbox"
           aria-expanded="false"
           aria-owns="list-24"
           class="v-input__slot"
      >
        <div class="v-select__slot">
          <label for="input-24"
                 class="v-label v-label--active theme--light"
                 style="left: 0px; position: absolute;"
          >
            Zichtbaarheid
          </label>
          <input id="input-24"
                 type="text"
                 autocomplete="off"
          >
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-menu-down theme--light"
              >
              </i>
            </div>
          </div>
          <input type="hidden"
                 value="4"
          >
        </div>
        <div class="v-menu">
          <div class="v-menu__content theme--light v-autocomplete__content"
               style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          >
          </div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper">
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-select v-autocomplete">
    <div class="v-input__control">
      <div role="combobox"
           aria-haspopup="listbox"
           aria-expanded="false"
           aria-owns="list-30"
           class="v-input__slot"
      >
        <div class="v-select__slot">
          <label for="input-30"
                 class="v-label v-label--active theme--light"
                 style="left: 0px; position: absolute;"
          >
            Hoofdcategorie
          </label>
          <input id="input-30"
                 type="text"
                 autocomplete="off"
          >
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-menu-down theme--light"
              >
              </i>
            </div>
          </div>
          <input type="hidden"
                 value="9"
          >
        </div>
        <div class="v-menu">
          <div class="v-menu__content theme--light v-autocomplete__content"
               style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          >
          </div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper">
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input theme--light v-text-field v-text-field--is-booted v-select v-autocomplete">
    <div class="v-input__control">
      <div role="combobox"
           aria-haspopup="listbox"
           aria-expanded="false"
           aria-owns="list-36"
           class="v-input__slot"
      >
        <div class="v-select__slot">
          <label for="input-36"
                 class="v-label theme--light"
                 style="left: 0px; position: absolute;"
          >
            Subcategorie
          </label>
          <input id="input-36"
                 type="text"
                 autocomplete="off"
          >
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-menu-down theme--light"
              >
              </i>
            </div>
          </div>
          <input type="hidden">
        </div>
        <div class="v-menu">
          <div class="v-menu__content theme--light v-autocomplete__content"
               style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          >
          </div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper">
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GlobalSearch.vue matches ui when a subcategory is set 1`] = `
<div class="category-selector">
  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-select v-autocomplete">
    <div class="v-input__control">
      <div role="combobox"
           aria-haspopup="listbox"
           aria-expanded="false"
           aria-owns="list-79"
           class="v-input__slot"
      >
        <div class="v-select__slot">
          <label for="input-79"
                 class="v-label v-label--active theme--light"
                 style="left: 0px; position: absolute;"
          >
            Zichtbaarheid
          </label>
          <input id="input-79"
                 type="text"
                 autocomplete="off"
          >
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-menu-down theme--light"
              >
              </i>
            </div>
          </div>
          <input type="hidden"
                 value="4"
          >
        </div>
        <div class="v-menu">
          <div class="v-menu__content theme--light v-autocomplete__content"
               style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          >
          </div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper">
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-select v-autocomplete">
    <div class="v-input__control">
      <div role="combobox"
           aria-haspopup="listbox"
           aria-expanded="false"
           aria-owns="list-85"
           class="v-input__slot"
      >
        <div class="v-select__slot">
          <label for="input-85"
                 class="v-label v-label--active theme--light"
                 style="left: 0px; position: absolute;"
          >
            Hoofdcategorie
          </label>
          <input id="input-85"
                 type="text"
                 autocomplete="off"
          >
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-menu-down theme--light"
              >
              </i>
            </div>
          </div>
          <input type="hidden"
                 value="9"
          >
        </div>
        <div class="v-menu">
          <div class="v-menu__content theme--light v-autocomplete__content"
               style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          >
          </div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper">
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-select v-autocomplete">
    <div class="v-input__control">
      <div role="combobox"
           aria-haspopup="listbox"
           aria-expanded="false"
           aria-owns="list-91"
           class="v-input__slot"
      >
        <div class="v-select__slot">
          <label for="input-91"
                 class="v-label v-label--active theme--light"
                 style="left: 0px; position: absolute;"
          >
            Subcategorie
          </label>
          <input id="input-91"
                 type="text"
                 autocomplete="off"
          >
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--append">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-menu-down theme--light"
              >
              </i>
            </div>
          </div>
          <input type="hidden"
                 value="17"
          >
        </div>
        <div class="v-menu">
          <div class="v-menu__content theme--light v-autocomplete__content"
               style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"
          >
          </div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper">
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GlobalSearch.vue matches ui when using shallowMount 1`] = `
<div class="category-selector">
  <vautocomplete-stub errorcount="1"
                      errormessages
                      messages
                      rules="function (v) { return !!v || 'Zichtbaarheid is een verplicht veld'; }"
                      successmessages
                      appendicon="$dropdown"
                      backgroundcolor
                      label="Zichtbaarheid"
                      loaderheight="2"
                      clearicon="$clear"
                      type="text"
                      valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
                      nodatatext="$vuetify.noDataText"
                      items="[object Object],[object Object]"
                      itemcolor="primary"
                      itemdisabled="disabled"
                      itemtext="name"
                      itemvalue="id"
                      menuprops="[object Object]"
                      allowoverflow="true"
                      filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }"
                      value="4"
  >
  </vautocomplete-stub>
  <vautocomplete-stub errorcount="1"
                      errormessages
                      messages
                      rules="function (v) { return !!v || 'Hoofdcategorie is een verplicht veld'; }"
                      successmessages
                      value="9"
                      appendicon="$dropdown"
                      backgroundcolor
                      label="Hoofdcategorie"
                      loaderheight="2"
                      clearicon="$clear"
                      type="text"
                      valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
                      nodatatext="$vuetify.noDataText"
                      items="[object Object],[object Object]"
                      itemcolor="primary"
                      itemdisabled="disabled"
                      itemtext="name"
                      itemvalue="id"
                      menuprops="[object Object]"
                      allowoverflow="true"
                      filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }"
  >
  </vautocomplete-stub>
  <vautocomplete-stub errorcount="1"
                      errormessages
                      messages
                      rules="function (v) { return !!v || 'Subcategorie is een verplicht veld'; }"
                      successmessages
                      value="17"
                      appendicon="$dropdown"
                      backgroundcolor
                      label="Subcategorie"
                      loaderheight="2"
                      clearicon="$clear"
                      type="text"
                      valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
                      nodatatext="$vuetify.noDataText"
                      items="[object Object],[object Object],[object Object]"
                      itemcolor="primary"
                      itemdisabled="disabled"
                      itemtext="name"
                      itemvalue="id"
                      menuprops="[object Object]"
                      allowoverflow="true"
                      filter="function _default(item, queryText, itemText) {
        return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
      }"
  >
  </vautocomplete-stub>
</div>
`;
